### PR TITLE
MANTA-4731 rebalancer should expose a REST API

### PR DESCRIPTION
**Pull request by rjloura**
This is an initial whack at a REST API.  It assumes we have the job_status work as it exposes endpoints to get status.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -67,3 +67,7 @@ unescape = "0.1.0"
 [patch.crates-io]
 diesel = { git = 'https://github.com/diesel-rs/diesel' }
 diesel_derives = { git = "https://github.com/diesel-rs/diesel" }
+
+[[bin]]
+name = "server"
```
Comment (rjloura) [inline Cargo.toml]: rename to rebalancer-manager

File: Cargo.toml [hunk 2]
```diff
@@ -67,3 +67,7 @@ unescape = "0.1.0"
...
+path = "src/server.rs"
```
Comment (rjloura) [inline Cargo.toml]: manager.rs

File: src/server.rs [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+            }
+        });
+    }
+
+    build_simple_router(|route| {
+        route
+            .get("/jobs/:uuid")
+            .with_path_extractor::<GetJobParams>()
+            .to(get_job);
+        route.get("/jobs").to(list_jobs);
```
Comment (rjloura) [inline src/server.rs]: use `to_new_handler()` to make async

File: src/server.rs [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+
+                let from_shark = evac_payload.from_shark.clone();
+                let job_action =
+                    JobAction::Evacuate(Box::new(EvacuateJob::new(
+                        from_shark,
+                        &domain_name,
+                        &job_uuid,
+                        max_objects,
+                    )));
+
```
Comment (rjloura) [inline src/server.rs]: Consider refactoring this.

File: src/server.rs [hunk 5]
```diff
@@ -0,0 +1,409 @@
...
+    fn default() -> Self {
+        JobActionPayload::Evacuate(EvacuateJobPayload::default())
+    }
+}
+
+#[derive(Serialize, Deserialize, Default)]
+struct EvacuateJobPayload {
+    from_shark: MantaObjectShark,
+    domain_name: Option<String>,
+    max_objects: Option<String>,
```
Comment (rjloura) [inline src/server.rs]: can this change to u32?

File: src/server.rs [hunk 6]
```diff
@@ -0,0 +1,409 @@
...
+                        } else {
+                            Some(val)
+                        }
+                    }
+                    None => {
+                        Some(10) // Default
+                    }
+                };
+
+                let from_shark = evac_payload.from_shark.clone();
```
Comment (rjloura) [inline src/server.rs]: Is this clone necessary?

File: src/server.rs [hunk 7]
```diff
@@ -0,0 +1,409 @@
...
+                Ok(j) => j,
+                Err(_) => {
+                    // TODO Check error
+                    return;
+                }
+            };
+            // This blocks until the job is complete.  If the user want's to
+            // see the status of the job, they can issue a request to:
+            //      /jobs/<job uuid>
+            if let Err(e) = job.run() {
```
Comment (rjloura) [inline src/server.rs]: Test to ensure only a single job can be run at once.

File: src/server.rs [hunk 8]
```diff
@@ -0,0 +1,409 @@
+#[macro_use]
+extern crate gotham_derive;
+
+#[macro_use]
+extern crate serde_derive;
+
+use remora::{info, log_impl, util, warn};
```
Comment (rjloura) [inline src/server.rs]: Is `log_impl` required?

File: src/server.rs [hunk 9]
```diff
@@ -0,0 +1,409 @@
...
+
+    fn unit_test_init() {
+        let mut init = INITIALIZED.lock().unwrap();
+        if *init {
+            return;
+        }
+
+        thread::spawn(move || {
+            let _guard = util::init_global_logger();
+            loop {}
```
Comment (rjloura) [inline src/server.rs]: put a sleep in here.

File: src/server.rs [hunk 10]
```diff
@@ -0,0 +1,409 @@
...
+        });
+
+        *init = true;
+    }
+
+    #[test]
+    fn basic() {
+        unit_test_init();
+        let config = Config::parse_config(Some("src/config.json")).unwrap();
+        let test_server = TestServer::new(router(config)).unwrap();
```
Comment (rjloura) [inline src/server.rs]: consider making a global.

File: src/server.rs [hunk 11]
```diff
@@ -0,0 +1,409 @@
...
+        println!("Retun value: {:#?}", job_list);
+
+        let get_this_job = job_list.jobs[0].clone();
+        let get_job_uri =
+            format!("http://localhost:8888/jobs/{}", get_this_job);
+
+        let response = test_server.client().get(get_job_uri).perform().unwrap();
+
+        let ret = response.read_body().unwrap();
+        let pretty_response: JobStatus = serde_json::from_slice(&ret).unwrap();
```
Comment (rjloura) [inline src/server.rs]: drop some asserts in here.

File: src/server.rs (hunk 3 above)
Comment (rjloura) [reply src/server.rs]: I found the behavior was the same with `to_new_handler()` so I implemented a futures based approach manually.

File: src/server.rs (hunk 4 above)
Comment (rjloura) [reply src/server.rs]: Cleaned this up some.

File: src/server.rs (hunk 5 above)
Comment (rjloura) [reply src/server.rs]: changed.

File: src/server.rs (hunk 8 above)
Comment (rjloura) [reply src/server.rs]: When `log_impl` is removed I see the following error:

> error: cannot find macro `log_impl!` in this scope
  --> src/manager.rs:48:5
   |
48 |     warn!("{}", msg);
   |     ^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)
```

File: src/server.rs (hunk 10 above)
Comment (rjloura) [reply src/server.rs]: Not currently having any issues with spawning multiple TestServers.

File: src/manager.rs [hunk 12]
```diff
@@ -162,7 +189,8 @@ impl Default for JobActionPayload {
 struct EvacuateJobPayload {
     from_shark: MantaObjectShark,
     domain_name: Option<String>,
-    max_objects: Option<String>,
+    //max_objects: Option<String>,
```
Comment (rhb2) [inline src/manager.rs]: Not necessary?

Comment (rjloura) [reply src/manager.rs]: good catch.

**Pull request merged**